apollo-federation/src/schema/schema_upgrader.rs

```diff
@@ -0,0 +1,174 @@
+use apollo_compiler::collections::HashMap;
+use apollo_compiler::Name;
+
+use super::TypeDefinitionPosition;
+use crate::error::FederationError;
+use crate::schema::SubgraphMetadata;
+use crate::ValidFederationSchema;
+use crate::ValidFederationSubgraph;
+use crate::ValidFederationSubgraphs;
+
+#[derive(Clone, Debug)]
+struct SchemaUpgrader<'a> {
+    schema: ValidFederationSchema,
+    name: String,
+    url: String,
+    subgraphs: &'a ValidFederationSubgraphs,
+    object_type_map: &'a HashMap<Name, HashMap<String, TypeInfo>>,
+}
+
+#[derive(Clone, Debug)]
+struct TypeInfo {
+    pos: TypeDefinitionPosition,
+    metadata: SubgraphMetadata,
+}
+
+pub(crate) fn upgrade_subgraphs_if_necessary(
+    subgraphs: ValidFederationSubgraphs,
+) -> Result<(), FederationError> {
+    let mut federation_subgraphs = ValidFederationSubgraphs::new();
+
+    // if all subgraphs are fed 2, there is no upgrade to be done
+    if subgraphs
+        .subgraphs
+        .iter()
+        .all(|(_name, subgraph)| subgraph.schema.is_fed_2())
+    {
+        return Ok(());
+    }
+
+    let mut object_type_map: HashMap<Name, HashMap<String, TypeInfo>> = Default::default();
+    for (_, subgraph) in subgraphs.subgraphs.iter() {
+        if let Some(subgraph_metadata) = subgraph.schema.subgraph_metadata() {
+            for pos in subgraph.schema.get_types() {
+                match pos {
+                    TypeDefinitionPosition::Object(_) | TypeDefinitionPosition::Interface(_) => {
```

Comment:
Since we aren't using any of the inner data of the variants and ignoring all other variants, I would use
```
if matches!(pos, TypeDefinitionPosition::Object(_) | TypeDefinitionPosition::Interface(_)) {
```
This cleans things up just a bit and you can remove the null branch.